pallets/farming/src/gauge.rs

```diff
@@ -0,0 +1,426 @@
+{
+    pub fn create_gauge_pool(
+        pool_id: PoolId,
+        pool_info: &mut PoolInfo<BalanceOf<T>, CurrencyIdOf<T>, AccountIdOf<T>, BlockNumberFor<T>>,
+        gauge_token: CurrencyIdOf<T>,
+        gauge_basic_rewards: BTreeMap<CurrencyIdOf<T>, BalanceOf<T>>,
+        max_block: BlockNumberFor<T>,
+    ) -> DispatchResult {
+        let gid = Self::gauge_pool_next_id();
+        pool_info.gauge = Some(gid);
+        let current_block_number = frame_system::Pallet::<T>::block_number();
+        let gauge_pool_info = GaugePoolInfo::new(
+            pool_id,
+            gauge_token,
+            pool_info.keeper.clone(),
+            pool_info.reward_issuer.clone(),
+            gauge_basic_rewards,
+            max_block,
+            current_block_number,
+        );
+
+        GaugePoolInfos::<T>::insert(gid, &gauge_pool_info);
+        GaugePoolNextId::<T>::mutate(|id| -> DispatchResult {
+            *id = id
+                .checked_add(One::one())
+                .ok_or(ArithmeticError::Overflow)?;
+            Ok(())
+        })?;
+        Ok(())
+    }
+
+    pub fn gauge_add(
+        who: &AccountIdOf<T>,
+        gid: PoolId,
+        gauge_value: BalanceOf<T>,
+        gauge_block: BlockNumberFor<T>,
+    ) -> DispatchResult {
+        GaugePoolInfos::<T>::mutate(gid, |gauge_pool_info_old| -> DispatchResult {
+            if let Some(mut gauge_pool_info) = gauge_pool_info_old.take() {
+                let mut gauge_info =
+                    GaugeInfos::<T>::get(gid, who).unwrap_or_else(|| GaugeInfo::new(who.clone()));
+
+                Self::process_gauge_add(
+                    &mut gauge_pool_info,
+                    &mut gauge_info,
+                    gauge_value,
+                    gauge_block,
+                )?;
+
+                T::MultiCurrency::transfer(
+                    gauge_pool_info.token,
+                    who,
+                    &gauge_pool_info.keeper,
+                    gauge_value,
+                )?;
+
+                GaugeInfos::<T>::insert(gid, who, gauge_info);
+                *gauge_pool_info_old = Some(gauge_pool_info);
+                Ok(())
+            } else {
+                Err(Error::<T>::GaugePoolNotExist)?
+            }
+        })?;
+        Ok(())
+    }
+
+    fn process_gauge_add(
+        gauge_pool_info: &mut GaugePoolInfoOf<T>,
+        gauge_info: &mut GaugeInfoOf<T>,
+        gauge_value: BalanceOf<T>,
+        gauge_block: BlockNumberFor<T>,
+    ) -> DispatchResult {
+        let current_block_number = frame_system::Pallet::<T>::block_number();
+        gauge_pool_info.gauge_last_block = current_block_number;
+        gauge_pool_info.gauge_amount = gauge_pool_info
+            .gauge_amount
+            .checked_add(&gauge_value)
+            .ok_or(ArithmeticError::Overflow)?;
+
+        ensure!(
+            gauge_info.gauge_stop_block >= current_block_number
+                || gauge_info.gauge_stop_block == Default::default(),
```

Comment:
when initialize a new gauge info, `gauge_stop_block` value is Default. the `process_gauge_add()` will update to either current block number(if gauge_amount=0) or increment `gauge_block` based on old value.
The `LastGaugeNotClaim` error happens when `gauge_stop_block` < current block number.
Because we allow process_gauge_add in initial case. so we need to open gate for `gauge_stop_block` is default.